rust/connlib/libs/common/src/session.rs

```diff
@@ -31,9 +31,9 @@ pub trait ControlSession<T, U> {
 /// A session is the entry-point for connlib, maintains the runtime and the tunnel.
 ///
 /// A session is created using [Session::connect], then to stop a session we use [Session::disconnect].
-pub struct Session<T, U, V, R, M> {
+pub struct Session<T, U, V, R, M, CB: Callbacks + 'static> {
```

Comment:
Having trait bounds on structs can become really cumbersome because they are quite infectious to the rest of the codebase. Unless you are referencing associated types (`Callbacks` doesn't have any from what I can see?) somewhere, I wouldn't do it.

Traits are about behaviour and behaviour is specified on `impl` blocks. `struct` declarations should be for layout only.

It is weird initially to have a type-parameter to be "too open" for what you want users to use it but in reality, they can't do anything with a "wrongly" instantiated type anyway if you have the correct bounds on your `impl` blocks which is where you need them anyway to actually implement your stuff and call the trait functions.

Reply:
> I understand your point, but I don't see the merit in `Session` having more degrees of freedom than we actually want or use.

Yeah, whether or not repeating the trait bounds is annoying depends a lot on how many layers of nesting you've got.

If your trait is object-safe, another way would be to use a trait-object instead of a type-parameter. That is equally expressive from a readability PoV but results in simpler type definitions.

Reply:
I'd rather not introduce trait-objects if we don't need to, that  would indicate some kind of runtime polymorphism that we don't actually need.

Reply:
> The fact that Rust requires trait bounds to be repeated does not change the fact that this type definition is my explicit intention. While your proposal may reduce visual noise, it increases conceptual/mental-model noise by introducing a degenerate case for Session types.

I agree with this but then we might want to remove the `'static` bound? That's just an implementation detail that we might even want to change in the future.